### ISSUE TITLE
request: ability to automatically generate tf_serving_warmup_requests files via TFX components

### ISSUE THREAD

**Issue by mattnworb** [tensorflow/tfx#2497]
Today we use TFX components (v0.15)  in Kubeflow pipelines to train and build SavedModels, which we then deploy tensorflow-serving. We have found that with tensorflow-serving, [warming-up the model](https://www.tensorflow.org/tfx/serving/saved_model_warmup) is very important to avoid excessive latency in the first few calls after the server starts up, since otherwise those calls will pay the cost of Tensorflow lazily-loading the graph within the model.

To encourage good practices with our internal teams that use TFX and Kubeflow, we wrote an extension of the `tensorflow_estimator.python.estimator.exporter.Exporter` Python class that generates the `tf_serving_warmup_requests` file by default based on supplied pre-transformed training examples, and then serializes it to the expected location in the SavedModel on disk. It was necessary to customize the training executor to support passing pre-transformed examples.

Pipelines then use this custom Exporter in their pipeline like

```python
# TFX invokes this function in our training module 
def trainer_fn(hparams, schema):
    
    # our custom class:
    exporter = ExporterWithWarmup(raw_train_files,  num_warmup_requests=100, ...)

   # our exporter is plugged into the EvalSpec
    eval_spec = tf.estimator.EvalSpec(
        eval_input_fn, exporters=[exporter], ...
    )

    return {
        "estimator": ...,
        "train_spec": ...,
        "eval_spec": eval_spec,
        "eval_input_receiver_fn": ...,
    }
```

This works great for generating the warmup files with TFX 0.15, but we would prefer to reduce the amount of subclassing and extension of TFX components that we have to maintain, especially in light of moving soon to TF2 and Keras-style estimators.

Since SavedModel warm-up is an important thing to consider when using tensorflow-serving in production, would it be possible to add built-in support for generating tf_serving_warmup_requests when producing a SavedModel, to avoid the need for writing custom code to do so?

**Comment by rclough** [comment]
Something also to consider is that by including raw/re-transformed examples with the model, this might be leaking sensitive/PII data by storing it with a model which is usually not conidered PII even if trained on such data. A possible solution might be to have something to generate fake warmup requests somehow, possibly based on the trainer's input signature?

/cc @zhitaoli

**Comment by 1025KB** [comment]
Hi, we are heading toward the GenericTrainer direction, in GenericExecutor the model save part is handled by user module in run_fn, as it's not necessary to be SavedModel format, GenericExecutor won't do any post run_fn processing.

A possible solution would be some utils that can be used in run_fn to do the warmup, wdyt?

**Comment by rclough** [comment]
This could work, potentially, although we have a number of (if not a majority of) customers models that are canned Estimators, and it's not clear when Keras will provide a similar functionality, so they will likely be using the Estimator trainer for quite some time, and it's hard to say if they will ever migrate off Estimator for deployed production models. For now we have a custom executor/component spec but as mentioned it's a bit of a pain to keep that in line with updates that happen in TFX.

For the generic use case, this is a nice step in the right direction, but not as nice as happening "for free" (or as default behavior). But I also understand how the `GenericTrainer` complicates things (or gives us the opportunity to push the onus onto end-users). If the behavior could somehow be coupled with using SavedModel format, that would be ideal - otherwise we are enforcing the use of these utils "by convention", which is a bit tenuous.

To explain why I focus so much on trying to have this behavior "for free" or by default - 

If you can imagine from an end user/customer's perspective (think something between a data scientist and a DS/MLE hybrid), they are mostly concerned with training a model, and model metric performance, etc when writing their `run_fn`. When it comes to serving the model, it's an arcane bit of info that you need to create this warmup file in order to have the expected performance during serving. Many of our customers probably aren't going to be aware of the need for these added tools when setting up their pipelines. This can be helped, to some degree, with templates/examples and documentation, but not everyone follows these things thoroughly. 

This is compounded by the fact that our ML infrastructure group tries to abstract away the details of model serving for our customers so they can focus on their skillset (ML) and our infra group can focus on what it takes to serve a model in a performant way at scale. It would seem then that we should manage creating these warmup files as part of this abstracted serving framework, but this requires modifying artifacts after publishing, which breaks TFX conventions, and it also breaks some internal conventions we have with our internal data publishing framework, since the data needs to be written into the directory of the published model artifact. We would be open to enforcing this somehow in the "Pusher" layer of things if it fit into the artifact publishing conventions.

But it would seem then, that for the purposes of saving and serving SavedModels, everyone could benefit from having this warmup file generated as a part of creating a SavedModel - otherwise the default behavior results in poor performance in TF serving, which isn't a great experience for a framework that claims to be high performance. If it doesn't happen there, then something about the Pusher process should support it, IMO.

Sorry it's a lot of text to explain this, but the flexibility/safety VS ease-of-use and performance are conflicting goals, which gets complicated in real world uses. We are interested in making TFX both easy and performant for our end-users, as we agree with the goal in the original paper, which is to reduce the amount of time spent thinking about infrastructure, so our engineers can focus on ML.

**Comment by 1025KB** [comment]
Note that GenericTrainer can be used for Estimator too, here is an [example](https://github.com/tensorflow/tfx/blob/master/tfx/components/testdata/module_file/trainer_module.py#L358)

I understand the importance of default. What I mean is we can provide the default in the trainer's module file.
For example, provide a template module file (similar idea like [this](https://github.com/tensorflow/tfx/blob/5efc6a6a4e406071bd0f3c7c60cba38e0b9def2b/tfx/experimental/templates/taxi/README.md)) that contains e.g., serving_fn, input_fn, run_fn, thus for most user, they just need to provide the model code in [build_model_fn](https://github.com/tensorflow/tfx/blob/master/tfx/examples/iris/iris_utils_native_keras.py#L113)

For a certain framework (e.g., estimator or keras), in normal cases, code in module file would be similar but we can't guarantee it's always the same, thus I feel template would be the solution to provide default than hardcode it in the component.

**Comment by rclough** [comment]
I understand what you mean by providing it as a template. We do this internally already. It would be helpful to have an official solution, but in my opinion, Tensorflow/TFX could, and would benefit from doing better than that. 

We have found it to be common common that end-users, not as familiar with TFX, delete code that they do not understand the function of, especially for something like warmup files, which do not directly impact training. This is what I mean when I say that "enforcing by convention" is not an ideal solution for something that can impact the performance of the model.

It seems to me that since we cannot assume the output format of the model, Tensorflow/Serving/TFX should either adopt a preferred serving model format that does not contain the performance issues SavedModel has without a warmup file, OR creating a SavedModel should, by default, generate a warmup file. In either case, maybe it would be better I open an issue with Tensorflow?

Alternatively, as I mentioned before, maybe it makes sense to move this logic into pusher. The only problem here is it's a bit more annoying to determine the model format at this stage.

**Comment by 1025KB** [comment]
I see, For file based Pusher, it's also not assume the model format type.

it's might be doable to add to trainer or pusher to check if the model is in certain format, we do some post process
or to be safe, user need to set certain flag to trigger the post process.

Let me check with internal to see how they handles warm up

**Comment by jiyongjung0** [comment]
Sorry for the late follow-up, but we added this feature starting from TFX 0.30.0 published last year.

https://github.com/tensorflow/tfx/blob/master/RELEASE.md#version-0300

Please take a look and re-open the issue if you have any further questions.
